design/back.compatibility.md

```diff
@@ -0,0 +1,79 @@
+I recommend option 1. 
+
+
+### Summary of Recommended Design
+We'll duplicate the files into a different subfolder per generation, add a generation-specific prefix to each file, and leave only the absolutely minimal set of files outside this folder. All of the code will be built into a single DLL.
+
+### Example
+For example, imagine the MRE SDK is at version 2.17.0. It would then support 3 generations of MREs: 1.0 (gen 1), 2.0 (gen 2), and 2.17.0 (pre-release for gen 3). 
+Any host app that includes this plugin will get permanent support for gen 1 and 2, and un-guaranteed prototype of gen3. If the host app later updates to SDK version 2.25.0, then gen 1 and gen 2 MREs run unchanged, but MREs built with gen3 version 2.17 to 2.24 may need to be updated 
+
+### Code to be shared across generations
+The more code is duplicated, the less chance of changes having side effects that spill into finalized generations, so the goal is to move as much as possible to be generational. There is *some* code that must be shared, though, in particular anything that interfaces with the main engine.
+
+#### Shared
+* Initial Connection code. Reason: until host connect to the MRE, it doesn't know the MRE's generation.
+* Code that the host app uses to initialize the MRE system and initialize a single MRE. Reason: host app doesn't need to care about which version each MRE is (MREAppsAPI)
+* Host app hooks related to input/user/attachments/createfromlibrary/video player. Reason: the host app generally doesn't care about which version an MRE is, and therefore should call into generation-agnostic hooks. Each generation-agnostic hook should then call into the generation-specific code with the appropriate parameters.
+* Logging/Debugging interface. While the logging functions are generational, the final formatted output should route back to the main engine through a generation-agnostic hook. Reason: the engine shouldn't worry ab
+* Network protocol/packet parsing (like newtonsoft.json). Reason: it's used to generate initial connection. Additional optimized protocols may be added later, which could probably be generational. 
+* RPC interface. Reason: This is used for MREs to talk directly to the main engine without going through the SDK, so it's host-app specific.
+
+#### Per generation
+* Anything that touches the scene graph - animation, physics, rendering, geometry, sound, video
+* Anything that touches assets - asset cache, shaders, procedural toolkit (including defaultmaterial)
+* Anything downloading files
+* External libraries that can change over time (including the gltf library)
+* Anything else, really, including utility functions, patching, payloads
+	
+### Steps
+1. Move almost all files into a gen1 folder (everything is considered generational unless explicitly chosen)
```

Comment:
Leveraging Visual Studio Automation may make this process manageable. They provide a pretty good C# API for interfacing with VS projects and files.

Reply:
Good tip